Make TabController animateTo() method work as expected

giantsol opened the pull request:
## Description

There are three parts that move in TabBar and TabBarView when we call tabController.animateTo():

1. TabBar's indicator
2. TabBar's scroll position (if it's scrollable)
3. TabBarView's scroll position

Currently, none of them properly follow the duration and curve given to tabController.animateTo(). I previously worked on a PR with only one fix among the three [here](https://github.com/flutter/flutter/pull/57799), but we've agreed it should rather be a breaking change and fix all of them together. So this PR is a re-work with all the three fixed along with a breaking change proposal.

**Before**

<p float="left">
  <img src="https://user-images.githubusercontent.com/4879766/85377492-23391a00-b574-11ea-881e-0d5048852bb9.gif" width="200" />
  <img src="https://user-images.githubusercontent.com/4879766/85384601-45369a80-b57c-11ea-9818-a172c9c91427.gif" width="200" />
</p>

Called `tabController.animateTo(0, duration: const Duration(seconds: 5))`, so the animation should finish after 5 seconds. However, you can see all the three elements above move much faster than that. Also, it doesn't look natural when I interrupt the animation by dragging the TabBarView.

**After**

<p float="left">
  <img src="https://user-images.githubusercontent.com/4879766/85377729-7743fe80-b574-11ea-8f80-93a56ee7e6e6.gif" width="200" />
  <img src="https://user-images.githubusercontent.com/4879766/85378119-15d05f80-b575-11ea-91cf-750c45f422ce.gif" width="200" />
</p>

TabController animation's duration and curve are ensured. Also, when user interrupts the animation, all the moving parts stops immediately altogether.

<details>
  <summary>Code</summary>

```dart
class Main extends StatefulWidget {
  Main({Key key}) : super(key: key);

  @override
  _MainState createState() => _MainState();
}

class _MainState extends State<Main> with SingleTickerProviderStateMixin {
  final List<Tab> _tabs = <Tab>[
    Tab(text: 'Tab 1'),
    Tab(text: 'Tab 2'),
    Tab(text: 'Tab 3'),
  ];

  TabController _tabController;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(vsync: this, length: _tabs.length);
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          bottom: TabBar(
            controller: _tabController,
            tabs: _tabs,
            isScrollable: true,
            labelPadding: const EdgeInsets.symmetric(horizontal: 56),
          ),
        ),
        body: TabBarView(
          controller: _tabController,
          children: _tabs.map((Tab tab) {
            return Center(
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Text(tab.text),
                  RaisedButton(
                    child: Text('animate to 0 in 5 seconds'),
                    onPressed: () {
                      _tabController.animateTo(0, duration: const Duration(seconds: 5));
                    },
                  ),
                ],
              ),
            );
          }).toList(),
        ),
      ),
    );
  }
}
```

</details>

## Related Issues

Fixes #55625

## Tests

Modified existing tests:
- `tabs_test.dart`: Overflowing RTL tab bar
- `tabs_test.dart`: TabBarView scrolls end close to a new page
- `tabs_test.dart`: TabBarView scrolls end close to a new page with custom physics

Added new tests:
- `tabs_test.dart`: TabController animation stops immediately when user starts dragging TabBarView
- `tabs_test.dart`: TabBar and TabBarView should follow TabController's animation duration and curve

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] Yes, this is a breaking change.
   - [x] I wrote a design doc: https://flutter.dev/go/respecting-tabcontroller-animateto-duration

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

giantsol committed:
Have TabBar indicator properly follow tabController animation value

giantsol committed:
Make TabController animation stop immediately when user starts dragging TabBarView

giantsol committed:
Make TabBar and TabBarView's paging/scrolling animation follow TabController's animation

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -205,6 +205,18 @@ class TabController extends ChangeNotifier {
+  ///
+  /// Defaults to [kTabScrollDuration].
+  Duration get lastDuration => _lastDuration;
+  Duration _lastDuration = kTabScrollDuration;
```
_lastDuration is not a good way to keep track an on going animation. Maybe use _onGoingAnimationDuration, and the property should be null when there is not animation on going. to prevent the previous animation accidentally leak to the next one

giantsol replied:
Sounds good. Will have a look at this.

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -282,7 +296,6 @@ class TabController extends ChangeNotifier {
   set offset(double value) {
     assert(value != null);
     assert(value >= -1.0 && value <= 1.0);
-    assert(!indexIsChanging);
```
The gif that shows interruption of the animateTo looks worse than i expected. Both the page and tab bar have a big screw. I think we should probably make it uninterruptible.

giantsol replied:
Ok, the only thing that changed in this PR was to make TabBar get interrupted. TabBarView was always getting interrupted because ```Scrollable``` widget inside TabBarView was cancelling the animation.
If we should make them all uninterruptible, I guess I should do something somewhere in Scrollable widget. I'll have a look at it.

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1197,6 +1198,7 @@ class _TabBarViewState extends State<TabBarView> {
   TabController _controller;
   PageController _pageController;
   List<Widget> _children;
+  List<Widget> _originalChildrenWithKey;
```
I presume all the changes regarding to _originalChildrenWithKey is due to the drag interruption?

giantsol replied:
Yep, that's right.

giantsol replied:
I remembered _originalChildrenWithKey is still needed while we don't support drag interruption. For instance, we're moving from 3rd tab to 1st tab with 10 seconds of duration, and if we **click** the 3rd tab, it still interrupts the animation. So we need to keep the original children in global.

chunhtai replied:
what would happen prior to this pr?

giantsol replied:
Say we're moving from 3rd tab to 1st tab with 10 seconds. That means ```_childenWithKey``` (not ```_originalChildrenWithKey```) are mingled in ```_warpToCurrentIndex()```. While the animation is running, if I click the 3rd tab, we enter ```_warpToCurrentIndex()``` again. Prior to this pr, what would happen is we mingle up the already mingled up ```_childrenWithKey```, so when all the animations finish, ```_childrenWithKey``` settles in non-original order.

giantsol replied:
Oops, I misunderstood your question. This was no problem prior to this pr because once animation is running inside ```_warpToCurrentIndex()```, it doesn't reenter that function until animation finishes (gets blocked in ```_handleTabControllerAnimationTick()```). However, now we're allowing it to reenter ```_warpToCurrentIndex()``` while animation is running (due to the changed condition in ```_handleTabControllerAnimationTick()```), so we need to keep the original children globally so they don't get messed up.

chunhtai replied:
If that is the case, why do we want to change the condition in `_handleTabControllerAnimationTick`? If we keep it the same, the clicking on tab will not interrupt the existing animation and it will not change the existing behavior of tab view.

giantsol replied:
The thing is, if we click the tab, ```TabBar``` always gets interrupted. If we keep the condition in ```_handleTabControllerAnimationTick``` the same, only ```TabBarView``` will not get interrupted and it will look like below:
<img src="https://user-images.githubusercontent.com/4879766/87698657-bc9fc880-c7ce-11ea-9fc1-500014f806c7.gif" width=150/>

However by changing the condition, we make both of them move to the same destination:
<img src="https://user-images.githubusercontent.com/4879766/87699161-6e3ef980-c7cf-11ea-9b41-abe906a7057a.gif" width=150/>

chunhtai replied:
ah I see, so the tabbarview will not get interrupt while tabbar will. what if we make the tabbar uninterruptible? (by do a condition check in _handleTap)

giantsol replied:
I can surely do that, but then user will have no way to interrupt the animation if it's taking long. Would that be okay?

chunhtai replied:
I believe that is the original intention but we only do it partially. I am open to make it interruptible, however, I can't think of a good way to make the animation to look ok(no weird jump or sudden content changes).

giantsol replied:
I also think there's no way to make the interrupted animation look ok as of now. Then I'll make TabBar click also uninterruptible and let you know.

chunhtai reviewed (commented):
I feel the way the pr tries to sync the duration and curve between tabcontroller and scrollcontroller is a little bit brittle. Especially the duration is depends on the start time. can we think of a way to make a more strong link between these two controllers instead of going through tabview?

giantsol commented:
@chunhtai 
I may need some help to keep this PR going. Unfortunately, I cannot find a way to make TabBarView uninterruptible. I've investigated for quite some time, but I still couldn't understand inner workings in ScrollActivity, ScrollableState etc :(
Making stronger link between controllers also doesn't look easy. The two controller values don't always match, and there's already lots of logic to deal with these edge cases. So the way I took was just to link the duration and curve without touching those logic. I think this wouldn't result in any serious timing issue you're concerned about because logic itself is the same as before.

Perhaps any help I can get? I don't want to leave this PR hanging around bothering you, but also feels too complicated to resolve on my own. Thanks.

chunhtai commented:
I think there are two ways we can do this.
1 Use ignorePointer in the tabview build method. Something like this should work
```dart
    return IgnorePointer(
      ignoring: _warpUnderwayCount > 0,
      child: NotificationListener<ScrollNotification>(
        onNotification: _handleScrollNotification,
        child: PageView(
          dragStartBehavior: widget.dragStartBehavior,
          controller: _pageController,
          physics: widget.physics == null
            ? _kTabBarViewPhysics.applyTo(const ClampingScrollPhysics())
            : _kTabBarViewPhysics.applyTo(widget.physics),
          children: _childrenWithKey,
        ),
      )
    );
```

2. Change the API of scroll controller to disable the scroll explicitly.

The 1 is probably a better approach if there is no unanticipated problem, which i can't think of any at the moment. However, ignorePointer has given me a lot of problem in the past. So I kinda hesitate to use it.  The 2 is the alternative but may require tons of works. I will probably go with 1 for now and see if there will be any issue.

giantsol commented:
@chunhtai 
Oh~ thanks for the hint! I'll give it a shot.

giantsol committed:
Make TabController animation not get interrupted by user dragging TabBarView

giantsol committed:
Change lastDuration and lastCurve to ongoingAnimationDuration and ongoingAnimationCurve

giantsol commented:
@chunhtai 
Sorry I did a force-push by mistake, but nothing in previous commits have changed. Added 2 new commits regarding code review :)

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1269,11 +1271,12 @@ class _TabBarViewState extends State<TabBarView> {
 
   void _handleTabControllerAnimationTick() {
-    if (_warpUnderwayCount > 0 || !_controller.indexIsChanging)
+    if (_warpUnderwayCount > 0 && _controller.index == _currentIndex)
```
is this still needed?

giantsol replied:
Yes, because while ```_warpToCurrentIndex()``` is running and user clicks another tab, we enter ```_handleTabControllerAnimationTick()``` and we need to allow it to go all the way to ```_warpToCurrentIndex()``` again. This is the case when ```_controller.index``` and ```_currentIndex``` are different.

This wasn't needed before because all the animations had same duration, so even without the condition I added, prior animation would finish before the new animation which is triggered by tab click, so it eventually reached ```_warpToCurrentIndex()``` again.

giantsol commented:
@chunhtai 
There's another problem. We can block user gesture during animation to prevent weird jumps happening, but what if we call ```tabController.animateTo(..)``` or ```tabController.index = ..``` during the animation? Weird jumps will happen again in that case unless we also block those calls during the animation, which doesn't sound good to me.

giantsol committed:
Make TabController animation not get interrupted by Tab click

giantsol committed:
Fix _ongoingAnimationDuration and _ongoingAnimationCurve not in sync when tabController.animteTo() is called in sequence with delay

giantsol commented:
@chunhtai 
I've made tab click not interrupt the animation, but kept the logic related with interruption (e.g. _originalChildrenWithKey) because they are still needed when ```tabController.animateTo()``` is called while animation is running. Now, user doesn't have a way to interrupt the animation, only code can interrupt it.

chunhtai reviewed (commented):
overall behavior seems fine, I left some comments

giantsol commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1291,36 +1297,41 @@ class _TabBarViewState extends State<TabBarView> {
     _pageController.jumpToPage(initialPage);
 
-    await _pageController.animateToPage(_currentIndex, duration: kTabScrollDuration, curve: Curves.ease);
+    await _pageController.animateToPage(_currentIndex, duration: _controller.ongoingAnimationDuration ?? kTabScrollDuration, curve: _controller.ongoingAnimationCurve ?? Curves.ease);
```
Yes, if we call ```tabController.animateTo(1, duration: Duration(seconds: 10))``` and after 1 second we call ```tabController.animateTo(3)``` it will come in here again while tabcontroller has been animating

chunhtai replied:
In that case i think we are fine, I more worry about that we are relying on tabController to notify this method as soon as it starts animating. If not, we are risking of animation may not be in sync because we use the duration instead of the remaining time of the animation in tab controller

giantsol commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1336,7 +1347,7 @@ class _TabBarViewState extends State<TabBarView> {
     if (notification is ScrollUpdateNotification && !_controller.indexIsChanging) {
       if ((_pageController.page - _controller.index).abs() > 1.0) {
-        _controller.index = _pageController.page.floor();
+        _controller.index = _pageController.page.round();
```
This changed was needed when we allowed user drag interruption during the tab controller animation. Since we now don't allow that to happen, we can revert it back to ```floor()```. However in a similar logic 5 lines below this we use ```round()```, and changing ```floor()``` to ```round()``` doesn't seem to have any side effect.

chunhtai replied:
line 1344 definitely has to be round() because we want to achieve the snapping effect. I think this is ok.

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -205,6 +205,18 @@ class TabController extends ChangeNotifier {
+  /// The duration given to [animateTo].
+  ///
+  /// Only valid while animation is running, null otherwise.
+  Duration get ongoingAnimationDuration => _ongoingAnimationDuration;
```
Instead of exposing this, which may only be useful when the animation first started (if the animation is already on the flight for a while, knowing the duration alone is not enough)

I am thinking maybe the tabcontroller should expose a new api to register a new callback, something like `onAnimationStarted`. The callback will take in the duration, curve and index. In the tabview, it will register a callback to also start animation on _pageController

giantsol replied:
Surely considerable. I was also not very fond of exposing ```ongoingAnimationDuration``` and ```ongoingAnimationCurve``` just for this purpose :( I'll try that out and see how it looks.
Perhaps I may also add ```onAnimationEnd``` along with ```onAnimationStart```?

giantsol committed:
Fix TabBar jag when user interrupts scroll

giantsol commented:
@chunhtai 
I tried implementing ```onAnimationStartedListener```, but it doesn't fit well in current architecture. It introduces some edge cases when used together with ```notifyListeners()``` in TabController, and it only gets messier if I try to somehow resolve these cases. I think keeping ```ongoingXX``` is a better choice for now, what do you think?

BTW, I found another bug while trying this, which was also present previous to this PR, and made a commit for that. Below was the bug:
<img src="https://user-images.githubusercontent.com/4879766/89183922-f7796d00-d5d2-11ea-95e3-1ac42a5c35fb.gif" width="300" />

It happens when you drag TabBarView, and before it settles to new index, you quickly touch the screen again.

chunhtai commented:
> @chunhtai
> I tried implementing `onAnimationStartedListener`, but it doesn't fit well in current architecture. It introduces some edge cases when used together with `notifyListeners()` in TabController, and it only gets messier if I try to somehow resolve these cases. I think keeping `ongoingXX` is a better choice for now, what do you think?
> 
> BTW, I found another bug while trying this, which was also present previous to this PR, and made a commit for that. Below was the bug:
> <img alt="" width="300" src="https://user-images.githubusercontent.com/4879766/89183922-f7796d00-d5d2-11ea-95e3-1ac42a5c35fb.gif">
> 
> It happens when you drag TabBarView, and before it settles to new index, you quickly touch the screen again.

nice find! That is definitely something we want to fix.

For the onAnimationStartedListener, Can you elaborate more on the corner cases you were facing?

giantsol committed:
Temporary commit to test AnimationStartedListener

giantsol commented:
@chunhtai 
It was hard to explain by word, so I made a [temporary commit](https://github.com/flutter/flutter/pull/60080/commits/1a36a95d939c42d82288f7315e51bf0152d2bfcc) showcasing AnimationStartedListener (note, this commit has bugs).
There were many things to consider. First, since we would want ```AnimationStartedListener``` to be called only when we call ```tabController.animateTo()```, we should keep using ```_handleTabControllerTick()``` and ```_handleTabControllerAnimationTick()``` to keep TabBar and TabBarView in sync with the controller. But this calls for bug because we call ```_notifyAnimationStartedListeners()``` right after ```notifyListeners()``` and TabBar will have already responded to ```_handleTabControllerTick()``` ignoring incoming ```_handleTabControllerAnimationStarted()```.  But we can't remove ```notifyListeners()``` because other people may have been using it.

Another approach is to call ```_notifyAnimationStartedListeners()``` not only when ```animateTo()``` is called but also when ```set index``` is called and let TabBar and TabBarView only listen to this callback (removing ```_handleTabControllerTick()```). But this also has problems because for instance, when user drags TabBarView and scroll settles, ```_handleScrollNotification()``` is called with ```ScrollEndNotification``` and inside, it calls ```_controller.index = ??```. This will call ```_notifyAnimationStartedListeners()``` again. Although it will have no impact to the UI (it is already at the target index), it will notify other listeners users could have registered.

I couldn't find appropriate way to mitigate all these problems :( Will be waiting for your comment!

chunhtai commented:
> _handleTabControllerTick

I am confused a little bit. There are three things here

1.  `tabcontroller.animationStartedListener`
2. `tabcontroller.listener`
3. `tabcontroller.animation.listener`

I took a look at your code, it looks like both TabBar and TabBarView are listening to all three of them, which causes this convoluted problem that one notification is conflicted with another.

My original idea was that the TabBar stays the same and listen to (2) and (3) so it should not have any problem with this change.

TabBarView only listens to (1) because this is the only thing it need in order to sync up the scroll controller with the tab controller. The only problem will be the case where you mentioned `user drags TabBarView and scroll settles` may cause a ballistic scroll and notify the animation start again, but we already have the issue before and we solved it by guarded it with `_warpUnderwayCount`. I wonder if we can do the same thing here?

I maybe missing some key problem here. Let me know what you think

giantsol commented:
Yes, we can guard ```TabBarView``` from responding to animation again when ballistic scroll ends, but we cannot stop ```TabController``` from notifying ```AnimationStartedListener``` because it will just be called when ```_controller.index = ?``` is called. If other users have added ```AnimationStartedListener``` to the tabController, they will be notified when ```ScrollEndNotification``` happens.
We could perhaps guard that by creating a new api like ```setIndex(0, notify = false)``` and call ```_controller.setIndex(0, notify = false)``` when ```ScroollEndNotification``` happens, but I don't think that's a good idea.

chunhtai commented:
> Yes, we can guard `TabBarView` from responding to animation again when ballistic scroll ends, but we cannot stop `TabController` from notifying `AnimationStartedListener` because it will just be called when `_controller.index = ?` is called. If other users have added `AnimationStartedListener` to the tabController, they will be notified when `ScrollEndNotification` happens.
> We could perhaps guard that by creating a new api like `setIndex(0, notify = false)` and call `_controller.setIndex(0, notify = false)` when `ScroollEndNotification` happens, but I don't think that's a good idea.

I realize we didnt do an equality check in index, most of time the clamping will continue the scroll until if scroll to the exact index and the ScroollEndNotification is called when `_controller.index == _pageController.page.round()` this will cause still trigger a meaningless animation. I think we should fix this.

If we fix that, I don't think we will have the issue you mentioned.

Also i didnt see the issue you mentioned in the gif 
 https://github.com/flutter/flutter/pull/60080#issuecomment-668004706. I ran my app against the master. Can you verify if you can still reproduce it on master branch

lazylazyllama commented:
I can confirm, that the issue in #60080 is reproducible in the current master branch.

giantsol commented:
@chunhtai 
Oh, okay. So you mean if we do some check before we call ```_controller.index = ?``` in ```ScrollEndNotification``` it will be okay? I don't think ```_controller.index == _pageController.page.round()``` will work there because we only update ```_controller.offset``` in ```ScrollUpdateNotification```, and updating offset doesn't update tabController's index value. ```_controller.index = ? ``` should be called when scroll ends to update tabController's index value and when that happens, AnimationStartedListener will be invoked.

Hmm.. I'm thinking perhaps we may be able to check ```_animationController.value == _index.toDouble()``` in tabController's ```_changeIndex()``` before we notify AnimationStartedListener.

@lazylazyllama 
Thanks!

giantsol committed:
Revert "Temporary commit to test AnimationStartedListener"

This reverts commit 1a36a95d939c42d82288f7315e51bf0152d2bfcc.

giantsol commented:
I'm working on a new commit to deal with edge cases. I'll let you know when it's done :)
BTW, when I run the test locally, it fails with below error:

![image](https://user-images.githubusercontent.com/4879766/89372586-81375080-d721-11ea-860e-5b5f03cf9db2.png)

I think it's related with this PR checks failing, but I haven't changed anything from SkiaGoldClient. Am I doing something wrong?

giantsol committed:
Merge remote-tracking branch 'upstream/master' into tabcontroller_animation

giantsol committed:
Use AnimationStartedListener instead of _ongoingAnimationXXX (need to remove comments)

giantsol commented:
@chunhtai 
I've reverted previous temporary commit and made a new working commit [here](https://github.com/flutter/flutter/pull/60080/commits/5f39b1eefeb004e73db16b4ea25a37990992288d). This commit deals with all edges cases I've found and seems to work nice. There are many edges cases to consider, and I've written some redundant comments to explain. Let me know what you think! Will delete those comments later.

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -224,13 +231,48 @@ class TabController extends ChangeNotifier {
+      // it was not called by ScrollEndNotification or
+      // index was changing before (so that TabBar and TabBarView will start a new animation with new target index.)
+      // This happens when controller.index = ? is called while animation was ongoing.
+      if (_animationController.value != _index.toDouble() || wasIndexChanging) {
```
I remember last time i check `ScrollEndNotification` is called controller.index = ? with the same value as controller.index. If we have the check on line 216, it should not have enter this block. Then this check is not necessary.

giantsol replied:
Same as above comment, we already had that check but still went all the way to notifying AnimationStartedListener, so we need this check.

chunhtai replied:
In that case we still need to notify the listener, if we think like a tabcontroller, there is no reason to block this because someone does ask us to change our index.

Either we fixes the tabview to not trigger this weird case or ignore this notify.

giantsol replied:
Ok, if we think ```AnimationStartedListener``` as ```IndexChangeStartedListener```, I guess notifying the listener does make sense.

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -224,13 +231,48 @@ class TabController extends ChangeNotifier {
+  }
+
+  /// Calls listener every time the animation starts by setting [index] or [animateTo].
+  /// When setting [index], duration is always [kTabScrollDuration] and
```
is this true? if you set the index directly, the duration should be null which should cause a warping effect on the tabbarview

chunhtai replied:
we should also mention the duration maybe null, and what does it mean to be null

giantsol replied:
Oh, the way I implemented it was to pass in ```kTabScrollDuration``` and ```Curves.ease``` to AnimationStartedListener when we're just setting the index:
![image](https://user-images.githubusercontent.com/4879766/89485627-2adb1800-d7dc-11ea-9db9-2fb611b94bbe.png)

So the duration and curve given to AnimationStartedListener are effectively NonNull. Would it be better we pass in null instead?

chunhtai replied:
I think we need to put null to correctly reflect what is actually happening inside tabcontroller. it is up to the observers to decide how to react to null duration(for example giving it kTabScrollDuration)

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -205,6 +206,8 @@ class TabController extends ChangeNotifier {
   /// [TabBarView.children]'s length.
   final int length;
 
+  final ObserverList<AnimationStartedListener> _animationStartedListeners = ObserverList<AnimationStartedListener>();
```
Since now the animation maybe called with null duration which is not really an animation, maybe we should find a better name here....

giantsol replied:
How about ```IndexChangeStartedListener```? I wonder if there's any naming standard in Flutter 🤔

chunhtai commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -313,6 +355,8 @@ class _TabControllerScope extends InheritedWidget {
   }
 }
 
+typedef AnimationStartedListener = void Function(Duration duration, Curve curve);
```
let's make it takes in a destination index as well.

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -914,6 +919,7 @@ class _TabBarState extends State<TabBar> {
     if (_controllerIsValid) {
       _controller.animation.removeListener(_handleTabControllerAnimationTick);
       _controller.removeListener(_handleTabControllerTick);
+      _controller.removeAnimationStartedListener(_handleTabControllerAnimationStarted);
```
Ah I didn't notice there is also a scroll controller in tab bar, but it seems like we can just use the _controller.removeAnimationStartedListener(_handleTabControllerAnimationStarted); alone will be enough?
Something like..
```
_handleTabControllerAnimationStarted(int destinationIndex, Duration duration, Curve curve) {
   if (duration == null)
      duration = kTabScrollDuration;
   _scrollToIndex(destinationIndex, duration, curve);
}

void _scrollToIndex(int destinationIndex, Duration duration, Curve curve) {
    final double offset = _tabCenteredScrollOffset(destinationIndex);
    _scrollController.animateTo(offset, duration: duration, curve: curve);
  }
```

I think this should be enough to keep tabcontroller and scroll controller in sync so we don't need _handleTabControllerTick and _handleTabControllerAnimationTick.
Let me know if i miss something.

giantsol replied:
First, we need ```_handleTabControllerAnimationTick()``` because tabbar scroll should keep in sync when user is dragging TabBarView. In that case, none of ```_handleTabControllerTick()``` or ```_handleTabControllerAnimationStarted()``` is called, so we need to do it in ```_handleTabControllerAnimationTick()```.

Second, we need ```_handleTabControllerTick()``` because TabController's ```_notifyAnimationStartedListeners()``` is not called when user drags and ballistic scroll ends. However, ```notifyListeners()``` is called, and we have to update TabBar's ```_currentIndex()``` when ballistic scroll ends, so we should do it in ```_handleTabControllerTick()```.

chunhtai replied:
I forgot about dragging yes, I agree _handleTabControllerAnimationTick is needed.

The _handleTabControllerTick though probably not needed if _notifyAnimationStartedListeners is called right? Although we need to figure out why the ballistic drag update does not drag the tabcontroller all the way to the index. Either case, _handleTabControllerAnimationStarted or _handleTabControllerAnimationTick should bring the scroll controller in tab bar to the right place?

giantsol replied:
Yep, we wouldn't need ```_handleTabControllerTick``` if we call ```_notifyAnimationStartedListeners```. I'll change this too.

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1265,28 +1283,29 @@ class _TabBarViewState extends State<TabBarView> {
+  void _handleTabControllerAnimationStarted(Duration duration, Curve curve) {
+    if (_warpUnderwayCount > 0 && _controller.index == _currentIndex)
       return; // This widget is driving the controller's animation.
 
```
we may need to take care of where duration is null

giantsol replied:
Yep, currently the duration and curve given are effectively NonNull because I passed in default values when not present. If we decide to pass in null instead, I'll take care of it here :)

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -1265,28 +1283,29 @@ class _TabBarViewState extends State<TabBarView> {
       return; // This widget is driving the controller's animation.
 
     if (_controller.index != _currentIndex) {
       _currentIndex = _controller.index;
```
Since we add the destination Index, let's use it here

giantsol commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -216,6 +219,10 @@ class TabController extends ChangeNotifier {
     _index = value;
```
![image](https://user-images.githubusercontent.com/4879766/89485453-cc159e80-d7db-11ea-933d-a1112e008164.png)

We were already doing it here! :)

giantsol committed:
Apply code review

- Change name AnimationStartedListener to IndexChangeStartedListener
- Remove redundant checks and call IndexChangeStartedListener whenever index changes
- Remove _handleTabControllerTick() from TabBar
- Remove redundant comments

giantsol commented:
@chunhtai 
Applied code review and removed redundant comments! Thanks!

giantsol committed:
Merge remote-tracking branch 'upstream/master' into tabcontroller_animation

chunhtai commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2677,6 +2691,142 @@ void main() {
+    expect(tabController.animation.value, 1.0);
+    expect(tabController.indexIsChanging, true);
+
+    await tester.pumpAndSettle();
```
Is this needed?

giantsol replied:
Oh I thought ```tester.pumpAndSettle()``` is needed to really end the test by finishing all the animations going on. Isn't this needed?

chunhtai replied:
It is ok to end the test in the middle of animation unless there are something specific you want to test.

giantsol replied:
Done!

chunhtai commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2677,6 +2691,142 @@ void main() {
     // because we tried to send a notification on dispose.
   });
+
+  testWidgets('TabController animation should not get interrupted when user drags TabBarView', (WidgetTester tester) async {
```
can you also add a test to make sure tap a tab can not interrupt the animation?

chunhtai replied:
also it will be perfect if you can add another test about the clipping bug you fixed as part of this pr

giantsol replied:
Done!

chunhtai reviewed (commented):
Overall looks great, mostly just nit picking.

giantsol committed:
Apply code review

- Change addIndexChangeStartedListener docs
- Add two tests

giantsol commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -227,12 +231,43 @@ class TabController extends ChangeNotifier {
+  }
+
+  /// Calls listener every time the index starts changing by setting [index] or [animateTo].
+  /// When setting [index], duration and curve are null.
```
Done!

giantsol committed:
Prevent _warpToCurrentIndex from being called redundantly when user drags TabBarView and settles to new index

giantsol commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -1000,18 +1000,22 @@ void main() {
 
     // Not close enough to switch to page 2
     pageController.jumpTo(500.0);
+    await tester.pumpAndSettle();
```
This was needed because of the changed condition in `_handleTabControllerIndexChangeStarted`. Previously, it always returned when `_warpUnderwayCount > 0`. But now we allow it to be interrupted when current warping index and new target index is different, so when `_controller.index = _pageController.page.round()` is called in `ScrollEndNotification`, `_warpToCurrentIndex` was called and it ran a meaningless `_pageController.animateToPage` once. This animation gets cancelled right away, so it doesn't do anything to UI but it kinda "dragged" a frame.

This is a bug because we actually don't need to run `_warpToCurrentIndex` in this case. So I changed the order of setting `_controller.index` and `_currentIndex` [here](https://github.com/flutter/flutter/pull/60080/commits/db8b31030d79446870a716718bf325c1ede9eec8) so that it wouldn't call unnecessary warping again.

giantsol committed:
Add _handleTabControllerTick again to fix TabBar selected color bug

giantsol commented:
I had to re-add `_handleTabControllerTick` in [this commit](https://github.com/flutter/flutter/pull/60080/commits/0303cb0c2c948166ee8eafab6f01229abbf6eb71) because of the bug that TabBar's selected color doesn't get updated properly: 
<img src="https://user-images.githubusercontent.com/4879766/89756860-9bf13700-db1e-11ea-99a9-5e86c02eea67.gif" width=200 />

I haven't looked deep but this bug is related with not calling `setState` in TabBar correctly. It happened after we started to use `_handleTabControllerIndexChangeStarted` instead of `_handleTabControllerTick` in TabBar. When we click a tab, tabController's animation runs. Previous to this PR, we called `setState` in TabBar everytime in `_handleTabControllerTick` i.e. before animation starts and **after** animation ends. However after we removed `_handleTabControllerTick` and called `setState` **only when animation starts**, this bug happened. It looks like we have to call setState again when animation ends, so I re-added it. Now it works like before :)

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -945,9 +949,9 @@ class _TabBarState extends State<TabBar> {
+  void _scrollToCurrentIndex(Duration duration, Curve curve) {
     final double offset = _tabCenteredScrollOffset(_currentIndex);
-    _scrollController.animateTo(offset, duration: kTabScrollDuration, curve: Curves.ease);
+    _scrollController.animateTo(offset, duration: duration ?? kTabScrollDuration, curve: curve ?? Curves.ease);
```
_scrollController.animateTo returns a future, we can add a promise after it finishes to set state so that we don't need to use `_handleTabControllerTick` for this.

giantsol replied:
Hm I've tried that, but it seems we have to call setState after `tabController.indexIsChanging` has become false. We call `_scrollController.animateTo` slightly earlier than tabController's `_animationController.animateTo`, so when I try to call setState after above future returns, it emits below error:
![image](https://user-images.githubusercontent.com/4879766/89852195-49208980-dbc9-11ea-8812-6c48eb5abf56.png)

It looks like it's related with using `_ChangeAnimation` and `_DragAnimation` in line 1074 of tabs.dart

chunhtai replied:
We need to rebuild right before the animation start to update _DragAnimation into _ChangeAnimation.
Adding the setState prior should does the trick
```
  void _handleTabControllerIndexChangeStarted(int index, Duration duration, Curve curve) {
    if (index != _currentIndex) {
      setState(() {
        _currentIndex = index;
      });
      if (widget.isScrollable)
        _scrollToCurrentIndex(duration, curve);
    }
  }
```

This works because animation always starts on the next frame so the setState will make it just in time to update the _ChangeAnimation at the end of this frame.

The tab highlight on the other hand is a bit weird, it looks like there is some weird issue around how we turn on and off the selected using _buildStyledTab around line 1084 to line 1093. I have not look into too deep what the problem is. but I bet there is something wrong inside how we lerp the text style in line 183

giantsol replied:
I applied the change you suggested, but the above tab text highlighting issue still persists :(

chunhtai replied:
Yeah, I think the highlight issue is probably related to how we do the animation lerping in _TabStyle. Sorry i have been a little persist on getting things sorted out. We have put in too much bandage fixes to a point that no one really know how it works exactly :(

It will be great if we can clean it up a little bit now.

chunhtai replied:
take the TabController._changeIndex for example. It makes zero sense that we need to notify the listener twice before and after animation ends in line 223 and line 229. The comment does not make sense either. However, that is required to update the _DragAnimation into _ChangeAnimation in TabBar and updates it back from _ChangeAnimation into _DragAnimation.

giantsol replied:
No problem :) I understand sorting things out is really important. I also agree it needs cleanup, but maybe that should be done apart from this PR? What do you think?

chunhtai replied:
yes, we don't need to clean up everything in this pr, i think the last thing we need to figure out to merge this pr is what cause the highlight issue, and if we can fix it without _handleTabControllerTick. I am ok with using _handleTabControllerTick if it is unavoidable as long as we know the reason behind it.

giantsol committed:
Remove unnecessary tester.pumpAndSettle();

giantsol committed:
Make tab highlighting work properly only with _handleTabControllerIndexChangeStarted

giantsol commented:
@chunhtai 
There are two ways we can take about the tab highlighting issue.

First way is using `_handleTabControllerTick` and rebuilding the tabs before and after the animation. Based on current logic, it is necessary we rebuild the tabs after the animation ends. When animation starts, only the previous tab and target tab get `_ChangeAnimation`, so only two of them lerp highlights as the animation moves on. When animation finishes and tabs are rebuilt (i.e. `_controller.indexIsChanging` has become `false`), we apply `_DragAnimation` to the current tab and its closest neighbors, so that they would lerp their highlights as user drags.

Second way is only using `_handleTabControllerIndexChangeStarted` and rebuilding the tabs **only before the animation**. The reason this approach caused a problem is that we rebuild the tabs only when `_controller.indexIsChanging` is `true`, so that the previous tab and the target tab gets `_ChangeAnimation` and just stay like that. Unlike the above case, the target tab and its neighbors don't get `_DragAnimation` anymore. What we can do to resolve this is setting `_ChangeAnimation` to the previous tab and `_DragAnimation` to the target tab and its neighbors all in one shot (i.e. not doing them separately based on `_controller.indexIsChanging`). That's what I did [here](https://github.com/flutter/flutter/pull/60080/commits/db8400a16ead15815ac71f75890e4a0e30d8aade). Everything works fine except one difference. Since we now apply `_DragAnimation` to the target tab's neighbors right away, while we animate to the target tab, the left or right tab will also be highlighted on its way like below:
<img src="https://user-images.githubusercontent.com/4879766/90034072-21314300-dcfb-11ea-9efb-5fe5a3a2e17b.gif" width=200 />

If you look closely, while we animate from Tab 3 -> Tab 1, Tab 2 also lerps highlighting. Previously, Tab 2 wouldn't have lerped  because only Tab 1 and Tab 3 would have `_ChangeAnimation`. This change is necessary if we shall only use `_handleTabControllerIndexChangeStarted` to rebuild the tabs, because now we have only one time to apply animations to the tabs. Tell me what you think about this :)

BTW, I did a small cleanup in this commit. I saw no reason we should have `_ChangeAnimation` and `_DragAnimation` separately (they both work on `_controller.animation.value`) so I merged them into `_TabChangeAnimation`.

chunhtai commented:
> @chunhtai
> There are two ways we can take about the tab highlighting issue.
> 
> First way is using `_handleTabControllerTick` and rebuilding the tabs before and after the animation. Based on current logic, it is necessary we rebuild the tabs after the animation ends. When animation starts, only the previous tab and target tab get `_ChangeAnimation`, so only two of them lerp highlights as the animation moves on. When animation finishes and tabs are rebuilt (i.e. `_controller.indexIsChanging` has become `false`), we apply `_DragAnimation` to the current tab and its closest neighbors, so that they would lerp their highlights as user drags.
> 
> Second way is only using `_handleTabControllerIndexChangeStarted` and rebuilding the tabs **only before the animation**. The reason this approach caused a problem is that we rebuild the tabs only when `_controller.indexIsChanging` is `true`, so that the previous tab and the target tab gets `_ChangeAnimation` and just stay like that. Unlike the above case, the target tab and its neighbors don't get `_DragAnimation` anymore. What we can do to resolve this is setting `_ChangeAnimation` to the previous tab and `_DragAnimation` to the target tab and its neighbors all in one shot (i.e. not doing them separately based on `_controller.indexIsChanging`). That's what I did [here](https://github.com/flutter/flutter/pull/60080/commits/db8400a16ead15815ac71f75890e4a0e30d8aade). Everything works fine except one difference. Since we now apply `_DragAnimation` to the target tab's neighbors right away, while we animate to the target tab, the left or right tab will also be highlighted on its way like below:
> <img alt="" width="200" src="https://user-images.githubusercontent.com/4879766/90034072-21314300-dcfb-11ea-9efb-5fe5a3a2e17b.gif">
> 
> If you look closely, while we animate from Tab 3 -> Tab 1, Tab 2 also lerps highlighting. Previously, Tab 2 wouldn't have lerped because only Tab 1 and Tab 3 would have `_ChangeAnimation`. This change is necessary if we shall only use `_handleTabControllerIndexChangeStarted` to rebuild the tabs, because now we have only one time to apply animations to the tabs. Tell me what you think about this :)
> 
> BTW, I did a small cleanup in this commit. I saw no reason we should have `_ChangeAnimation` and `_DragAnimation` separately (they both work on `_controller.animation.value`) so I merged them into `_TabChangeAnimation`.

So the issue is because the tabbar is not rebuilt before and after the animation.
I suggest we use an animationStatusListener. This achieve the same result as _handleTabTick
```
_controller.animation.addStatusListener(_handleTabControllerAnimationStatusChanged);
void _handleTabControllerAnimationStatusChanged(AnimationStatus status) {
    setState(() {
    });
  }
```

Although they looks the same, this is more programmatically correct because we create a hack to make _handleTabTick call before and after an animation.

giantsol committed:
Revert "Make tab highlighting work properly only with _handleTabControllerIndexChangeStarted"

This reverts commit db8400a16ead15815ac71f75890e4a0e30d8aade.

giantsol commented:
I tried your suggestion, but unfortunately there seems to be a timing issue :(
I get this error when I call `setState` from `_handleTabControllerAnimationStatusChanged`:
![image](https://user-images.githubusercontent.com/4879766/90209617-d01e6d80-de26-11ea-8b9b-013417dea9a6.png)

It's from an assertion in _DragAnimation.

chunhtai commented:
> I tried your suggestion, but unfortunately there seems to be a timing issue :(
> I get this error when I call `setState` from `_handleTabControllerAnimationStatusChanged`:
> ![image](https://user-images.githubusercontent.com/4879766/90209617-d01e6d80-de26-11ea-8b9b-013417dea9a6.png)
> 
> It's from an assertion in _DragAnimation.

It was working fine for me last time i tried, you shouldn't need to setState in _handleTabControllerAnimationStatusChanged if you add animation status listener. Do you have a code that you see this issue?

giantsol committed:
Use _handleTabControllerAnimationStatusChanged instead of _handleTabControllerTick

giantsol commented:
Sure, I pushed a commit using `_handleTabControllerAnimationStatusChanged` so you could test with it. If you run below code with my current branch you can see the error:
```dart
void main() {
  runApp(Main());
}

class Main extends StatefulWidget {
  Main({Key key}) : super(key: key);

  @override
  _MainState createState() => _MainState();
}

class _MainState extends State<Main> with SingleTickerProviderStateMixin {
  final List<Tab> _tabs = <Tab>[
    Tab(text: 'Tab 1'),
    Tab(text: 'Tab 2'),
    Tab(text: 'Tab 3'),
    Tab(text: 'Tab 4'),
  ];

  TabController _tabController;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(vsync: this, length: _tabs.length);
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          bottom: TabBar(
            controller: _tabController,
            tabs: _tabs,
            isScrollable: true,
            labelPadding: const EdgeInsets.symmetric(horizontal: 56),
          ),
        ),
        body: TabBarView(
          controller: _tabController,
          children: _tabs.map((Tab tab) {
            return Center(
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Text(tab.text),
                  RaisedButton(
                    child: Text('animate to 0 in 5 seconds'),
                    onPressed: () async {
                      _tabController.animateTo(0, duration: const Duration(seconds: 5));
                    },
                  ),
                ],
              ),
            );
          }).toList(),
        ),
      ),
    );
  }
}
```

Try clicking Tab 3, then swipe to Tab 2 and then click Tab 3 again. If you use swipe and click a tab a few times you will see the error :)

chunhtai commented:
Thanks for the repro, I think you can fix this with
```dart
void _handleTabControllerAnimationStatusChanged(AnimationStatus status) {
    switch(status) {
      case AnimationStatus.completed:
        setState(() {
          // Rebuild the tabs after a (potentially animated) index change has
          // completed to update the tab animation back to _DragAnimation.
        });
        return;
      case AnimationStatus.reverse:
      case AnimationStatus.dismissed:
      case AnimationStatus.forward:
        return;
    }
  }

  void _handleTabControllerIndexChangeStarted(int index, Duration duration, Curve curve) {
    if (index != _currentIndex) {
      setState(() {
        // Updates the the tab animation to _ChangeAnimation.
      });
      _currentIndex = index;
      if (widget.isScrollable)
        _scrollToCurrentIndex(duration, curve);
    }
  }
```

giantsol committed:
Fix _handleTabControllerAnimationStatusChanged logic to work properly

giantsol commented:
@chunhtai 
I've applied the change you suggested and checked it working properly!
I see no specific problems with current implementation :)

chunhtai commented:
Hi @giantsol, sorry for the late reply. I had a discussion with the teams, and the general consent is we should not rely on duration and curve of the tabcontroller to drive scrollcontroller of both tabbar and tabbarview because the duration and curve can change mid flight. That means the IndexChangeStarted is not going to work. We will have to write some special logic to be able to drive the scrolling using the tab.animation. Fortunately, the tabbar and tabbarview owns the page controller and scroll controller, we should be able to write an private class of these two controllers to accept an animation value to listen to and creates their own scrollposition that will animate based on the animation value.

The full discussion can be found in our discord flutter-framework channel, I will post the transcript here too https://gist.github.com/chunhtai/0574c29e79b26e54e42fdea39063aa50

let me know if you have any question.

giantsol commented:
@chunhtai 
Thanks, I read the messages you had with Hixie. Of course we should follow what the team agreed on, but it looks like it would require another big change. I presume all the logic related with whether we should follow `tabController.animation.value` or not should all move into the new private class you mentioned. 

However, I doubt I could work on the new change you're suggesting. I guess it'd require a lot more time, but I just can't afford that. Stepping back, **this PR's main goal was to fix the bug that tabController.animateTo wasn't working properly, and it actually does fix that**. So I wanted to suggest we complete this PR and make it a new task for the new path you mentioned, but the problem is we'd be exposing the new `IndexChangeStartedListener` api which could soon be deprecated (if anyone starts working on a big change). There seems to be no _package private_ stuff in dart, so I guess there's no way we could stop exposing `IndexChangeStartedListener` :(

What do you think? There's been a lot going on in this PR. I guess we'd have to either find a way to merge this and then move on, or close this and let someone else work on it anew. Thanks.

chunhtai commented:
Hi @giantsol 

We should not merge a temporary solution, but I think we can reuse part of this pr because there are a lot of other fixes in this PR that are parallel from the indexchangestart listener. I am happy to keep working through the new solution with you or have someone else take over it. Let me know what you think. We are very thankful for your contribution either way.

giantsol commented:
@chunhtai 
Then I guess I should hand this over to someone else :) I studied a lot about TabController, TabBar and TabBarView along the way, and I'd be happy to help if anyone would need help on this. I would work on some small bug fixes we've encountered here though! I'll be sending new PRs for them soon.

Thank you so much for helping me out this far. It was my pleasure :)
